src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+LedgerState::getLoadBestOfferContext(Asset const& selling, Asset const& buying)
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter == mLoadBestOfferContext.end())
+    {
+        auto res = mLoadBestOfferContext.insert(std::make_pair(
+            assetPair, LoadBestOfferContext(selling, buying, *this)));
+        iter = res.first;
+    }
+    return iter->second;
+}
+
+void
+LedgerState::invalidateLoadBestOfferContext(Asset const& selling,
+                                            Asset const& buying)
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter != mLoadBestOfferContext.end())
+    {
+        mLoadBestOfferContext.erase(iter);
+    }
+}
+
+std::vector<LedgerState::StateEntry>
+LedgerState::getOffers(Asset const& selling, Asset const& buying)
+{
+    std::vector<StateEntry> offers;
+    std::set<LedgerKey> seen;
+    getOffers(selling, buying, offers, seen);
+    return offers;
+}
+
+void
+LedgerState::getOffers(Asset const& selling, Asset const& buying,
+                       std::vector<StateEntry>& offers,
+                       std::set<LedgerKey>& seen)
+{
+    for (auto const& state : mState)
+    {
+        if (state.first.type() != OFFER)
+        {
+            continue;
+        }
+        else if (!seen.insert(state.first).second)
+        {
+            continue;
+        }
+        else if (!state.second->ignoreInvalid().entry())
+        {
+            continue;
+        }
+        assert(!state.second->valid());
+
+        // Note: Can't compare Assets with !=
+        auto const& offer = state.second->ignoreInvalid().entry()->data.offer();
+        if (!(selling == offer.selling) || !(buying == offer.buying))
+        {
+            continue;
+        }
+        offers.push_back(state.second);
+    }
+
+    if (mParent)
+    {
+        mParent->getOffers(selling, buying, offers, seen);
+    }
+}
+
+LedgerState::LoadBestOfferContext::LoadBestOfferContext(
+    Asset const& selling, Asset const& buying, LedgerState& ledgerState)
+    : mSelling(selling)
+    , mBuying(buying)
+    , mLedgerState(ledgerState)
+    , mTop(nullptr)
+    , mInMemory([](StateEntry const& lhs,
+                   StateEntry const& rhs) { return compareOffers(rhs, lhs); },
+                mLedgerState.getOffers(selling, buying))
+    , mOffersLoadedFromDatabase(0)
+{
+}
+
+void
+LedgerState::LoadBestOfferContext::loadFromDatabaseIfNecessary()
+{
+    if (mFromDatabase.empty())
+    {
+        LedgerState* root = &mLedgerState;
+        while (root->mParent != nullptr)
+        {
+            root = root->mParent;
+        }
+        mFromDatabase = root->loadBestOffersFromDatabase(
+            5, mOffersLoadedFromDatabase, mSelling, mBuying);
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
```

Comment:
I think maybe "popNextBestOffer" would be a better name, as it will be called multiple times and each time it should return different value. load preffix suggest that it should return the same value each time.

Reply:
The semantics here are actually more like "loadBestOffer" then "popNextBestOffer". `loadBestOffer` has analogous functionality to `std::priority_queue<...>::top()`. For example, the following test case should succeed:
```
Asset a1, a2 = ...;
LedgerState ls(...);
auto ler1 = ls.loadBestOffer(a1, a2);
auto le1 = *ler1->entry();
ler1->invalidate();
auto ler2 = ls.loadBestOffer(a1, a2);
auto le2 = *ler2->entry();
REQUIRE(le1 == le2);
```
If you want it to behave like "popNextBestOffer" than you would have to use `LedgerEntryReference::erase` like:
```
Asset a1, a2 = ...;
LedgerState ls(...);
auto ler1 = ls.loadBestOffer(a1, a2);
auto le1 = *ler1->entry();
ler1->erase(); // Now we have popped
ler1->invalidate();
auto ler2 = ls.loadBestOffer(a1, a2);
auto le2 = *ler2->entry();
REQUIRE(le1 != le2);
```
This functionality is natural to me, since you do not want to remove the best offer from the database every time you load it.